pvlib/pvsystem.py

```diff
@@ -1860,51 +1880,111 @@ def v_from_i(resistance_shunt, resistance_series, nNsVth, current,
-
-    # Eqn. 3 in Jain and Kapoor, 2004
-    V = -I*(Rs + Rsh) + IL*Rsh - nNsVth*lambertwterm + I0*Rsh
-
-    return V
+    # asarray turns Series into arrays so that we don't have to worry
+    #  about multidimensional broadcasting failing
+    # Note that lambertw function doesn't support float128
+    Rsh = np.asarray(resistance_shunt, np.float64)
+    Rs = np.asarray(resistance_series, np.float64)
+    nNsVth = np.asarray(nNsVth, np.float64)
+    I = np.asarray(current, np.float64)
+    I0 = np.asarray(saturation_current, np.float64)
+    IL = np.asarray(photocurrent, np.float64)
+
+    # This transforms any ideal Rsh=np.inf into Gsh=0., which is generally
+    #  more numerically stable
+    Gsh = 1./Rsh
+
+    # Intitalize output V (including shape) by solving explicit model with
+    #   Gsh=0, multiplying by np.ones_like(Gsh) identity in order to also
+    #   capture shape of Gsh.
+    V = (nNsVth*np.log1p((IL - I)/I0) - I*Rs)*np.ones_like(Gsh)
+
+    # Record if inputs were all scalar
+    output_is_scalar = np.isscalar(V)
+
+    # Multiply by np.atleast_1d in order to convert scalars to arrays, because
+    #  we need to guarantee the ability to index arrays
+    V = np.atleast_1d(V)
+
+    # Expand Gsh input shape to match output V (if needed)
+    Gsh = Gsh*np.ones_like(V)
+
+    # Determine indices where Gsh>0 requires implicit model solution
+    idx = 0. < Gsh
+
+    # Only compute using LambertW if there are cases with Gsh>0
+    if np.any(idx):
+        # Expand remaining inputs to accomodate common indexing
+        Rs = Rs*np.ones_like(V)
+        nNsVth = nNsVth*np.ones_like(V)
+        I = I*np.ones_like(V)
+        I0 = I0*np.ones_like(V)
+        IL = IL*np.ones_like(V)
+
+        # LambertW argument, argW cannot be float128
+        argW = I0[idx] / (Gsh[idx]*nNsVth[idx]) * \
+            np.exp((-I[idx] + IL[idx] + I0[idx]) / (Gsh[idx]*nNsVth[idx]))
+
+        # lambertw typically returns complex value with zero imaginary part
+        lambertwterm = lambertw(argW).real
+
+        # Record indices where LambertW input overflowed output
+        idx_w = np.logical_not(np.isfinite(lambertwterm))
+
+        # Only re-compute LambertW if it overflowed
+        if np.any(idx_w):
+            # Calculate using log(argW) in case argW is really big
+            logargW = (np.log(I0[idx]) - np.log(Gsh[idx]) -
+                       np.log(nNsVth[idx]) +
+                       (-I[idx] + IL[idx] + I0[idx]) /
+                       (Gsh[idx] * nNsVth[idx]))[idx_w]
+
+            # Three iterations of Newton-Raphson method to solve
+            #  w+log(w)=logargW. The initial guess is w=logargW. Where direct
+            #  evaluation (above) results in NaN from overflow, 3 iterations
+            #  of Newton's method gives approximately 8 digits of precision.
+            w = logargW
+            for _ in range(0, 3):
+                w = w * (1 - np.log(w) + logargW) / (1 + w)
+            lambertwterm[idx_w] = w
```

Comment:
Looking at the coverage report, we never hit lines 1937-1949, even though one of the pre-existing unit tests (copied to the fixture in this PR) is supposed to cover this branch of the code. It may be that recasting in terms of Gsh helps avoid the overflow. @cwhanse Any ideas here?

Reply:
The unit test was from #226, I think.

Reply:
Not sure why the parameters from #225 don't cause entry into that code block.  They cause overflow on the np.exp() call.

```
def test_v_from_i_bigger():
    # 1000 W/m^2 on a Canadian Solar 220M with 20 C ambient temp
    # github issue 225
    output = pvsystem.v_from_i(190, 1.065, 2.89, 0, 7.05196029e-08, 10.491262)
'''

Reply:
So it appears that the Gsh recast helps the existing test's computation stay finite:

In[5]: pvlib.pvsystem.v_from_i(190, 1.065, 2.89, 0, 7.05196029e-08, 10.491262)
Using LambertW with argW=[  1.64103617e+294] and lambertwterm=[ 670.94665556]
Out[5]: 54.303958833791285

However, the original parameters from #225  serve as a good test, I think:

In[5]: pvlib.pvsystem.v_from_i(381.68, 1.065, 2.681527737715915, 0, 1.8739027472625636e-09, 5.1366949999999996)
Using LambertW with argW=[ inf] and lambertwterm=[ inf]
/Users/markcampanelli/Documents/Programming/pvlib-python/pvlib/pvsystem.py:1926: RuntimeWarning: overflow encountered in exp
  np.exp((-I[idx] + IL[idx] + I0[idx]) / (Gsh[idx]*nNsVth[idx]))
Out[4]: 58.19323124611128

Note that in followup work I might be interested in checking how accurate lambertw() is for such large inputs.

I will add this parameter set as a unit test and verify the coverage